vsintegration/src/FSharp.Editor/Navigation/GoToDefinitionService.fs

```diff
@@ -30,74 +34,372 @@ type internal FSharpNavigableItem(document: Document, textSpan: TextSpan) =
+    let findDeclarationOfSymbolAtRange
+        (targetDocument:Document, symbolRange:range, targetSource:SourceText, checker: FSharpChecker, projectInfoManager: ProjectInfoManager) =
+        findSymbolHelper (targetDocument, symbolRange, targetSource,true, checker, projectInfoManager) 
+
+
+    /// find the definition location (implementation file/.fs) of the target symbol
+    let findDefinitionOfSymbolAtRange
+        (targetDocument:Document, symbolRange:range, targetSourceText:SourceText, checker: FSharpChecker, projectInfoManager: ProjectInfoManager) =
+        findSymbolHelper (targetDocument, symbolRange, targetSourceText,false, checker, projectInfoManager)
+
+    
+    /// use the targetSymbol to find the first instance of its presence in the provided source file
+    let findSymbolDeclarationInFile
+        (targetSymbolUse:FSharpSymbolUse, filePath:string, source:string, checker:FSharpChecker, projectOptions:FSharpProjectOptions, fileVersion:int) = 
+        asyncMaybe {
+            let! (_parseResults, checkFileAnswer) = 
+                checker.ParseAndCheckFileInProject (filePath, fileVersion,source,projectOptions)|> liftAsync //(implDoc, projectOptions, allowStaleResults=true, sourceText=implSourceText)
+            match checkFileAnswer with 
+            | FSharpCheckFileAnswer.Aborted -> return! None
+            | FSharpCheckFileAnswer.Succeeded checkFileResults ->
+                let! symbolUses = checkFileResults.GetUsesOfSymbolInFile targetSymbolUse.Symbol |> liftAsync
+                let! implSymbol  = symbolUses |> Array.tryHead 
+                return implSymbol.RangeAlternate
         }
+
+
+    let statusBarMessage (statusBar:IVsStatusbar) (msg:string) =
+        let (_,frozen) = statusBar.IsFrozen()
+        // unfreeze the status bar
+        if frozen <> 0 then statusBar.FreezeOutput 0 |> ignore
+        statusBar.SetText msg |> ignore
+        // freeze the status bar
+        statusBar.FreezeOutput 1 |> ignore
+
+    let clearStatusBar (statusBar:IVsStatusbar) =
+        // unfreeze the statusbar
+        statusBar.FreezeOutput 0 |> ignore  
+        statusBar.Clear () |> ignore
+
+
+    let clearStatusBarAfter (statusBar:IVsStatusbar) (timeoutMs:int) =
+        Async.Sleep timeoutMs
+        |> Async.map (fun _ -> clearStatusBar statusBar)
+        |> Async.Start
+
+
+open FSharpGoToDefinition
+
+[<Shared>]
+[<ExportLanguageService (typeof<IGoToDefinitionService>, FSharpCommonConstants.FSharpLanguageName)>]
+[<Export (typeof<FSharpGoToDefinitionService>)>]
+type internal FSharpGoToDefinitionService [<ImportingConstructor>]
+    (checkerProvider: FSharpCheckerProvider,
+     projectInfoManager: ProjectInfoManager,
+     [<ImportMany>] presenters: IEnumerable<INavigableItemsPresenter>) =
+
+    let serviceProvider =  ServiceProvider.GlobalProvider  
+    let statusBar = serviceProvider.GetService<SVsStatusbar,IVsStatusbar>()
+
+
+    let tryNavigateToItem (navigableItem:#INavigableItem option) =
```

Comment:
I think we should use a wait dialog for this - see how I did it in the project system. It's more inline with how VS does blocking operations (show a progress dialog after a second or so).

Also of the strings in this file need localising.